alicloud/resource_alicloud_oss_bucket.go

```diff
@@ -0,0 +1,703 @@
+					buf.WriteString(fmt.Sprintf("%s-", m["target_bucket"]))
+					buf.WriteString(fmt.Sprintf("%s-", m["target_prefix"]))
+					return hashcode.String(buf.String())
+				},
+				MaxItems: 1,
+			},
+
+			"logging_isenable": &schema.Schema{
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+
+			"referer_config": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"allow_empty": {
+							Type:     schema.TypeBool,
+							Optional: true,
+							Default:  true,
+						},
+						"referers": {
+							Type:     schema.TypeList,
+							Required: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+					},
+				},
+				MaxItems: 1,
+			},
+
+			"lifecycle_rule": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							Computed:     true,
+							ValidateFunc: validateOssBucketLifecycleRuleId,
+						},
+						"prefix": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"enabled": {
+							Type:     schema.TypeBool,
+							Required: true,
+						},
+						"expiration": {
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"date": {
+										Type:         schema.TypeString,
+										Optional:     true,
+										ValidateFunc: validateOssBucketDateTimestamp,
+									},
+									"days": {
+										Type:     schema.TypeInt,
+										Optional: true,
+									},
+								},
+							},
+						},
+					},
+				},
+				MaxItems: 1000,
+			},
+
+			"creation_date": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"extranet_endpoint": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"intranet_endpoint": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"location": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"owner": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"storage_class": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAlicloudOssBucketCreate(d *schema.ResourceData, meta interface{}) error {
+	ossconn := meta.(*AliyunClient).ossconn
+
+	bucket := d.Get("bucket").(string)
+	isExist, err := ossconn.IsBucketExist(bucket)
+	if err != nil {
+		return err
+	}
+	if isExist {
+		return fmt.Errorf("[ERROR] The specified bucket name: %#v is not available. The bucket namespace is shared by all users of the OSS system. Please select a different name and try again.", bucket)
+	}
+
+	log.Printf("[DEBUG] OSS bucket create: %#v, using endpoint: %#v", bucket, ossconn.Config.Endpoint)
+
+	retryErr := resource.Retry(3*time.Minute, func() *resource.RetryError {
```

Comment:
This retry logic seems unnecessary; if the `create` API call fails, we immediately exit the retry loop. If it does not return an error, the inner function that would be retried returns `nil`, which will then immediately exit the retry loop.